### PR TITLE
Fixes #37486 - Clean up BMC sessions by logging out

### PR DESCRIPTION
**Pull request by ekohl**
After using a BMC connection, this logs out of the API. This is needed for Redfish, or the host will run out of available sessions. For other BMCs it's a noop.

This is an untested alternative to https://github.com/theforeman/smart-proxy/pull/895. I'm not sure yet what's best, especially looking at the performance since I don't have Redfish hardware to test with. One advantage of this approach is that it doesn't patch the RedfishClient class. That gem also looks unmaintained these days, so getting a proper patch in to support choosing behavior will be hard.

### CODE REVIEW

**Comment by neomilium** [comment]
To help us to take the right decision, I made some benchmarks.

I wrote 5 scripts in shell which use `curl` to:

1. Read Redfish Root as guest (ie. no authentications)
2. Read Redfish Root using basic auth
3. Read Redfish Root with session auth (ie. login, root, logout)
4. Request 10 endpoints using session auth (ie. login, 10 requests, logout)
5. Request 10 endpoints using basic auth (ie. send basic auth on each request)

Here some results:

```
              user     system      total        real
curl on root without login:  0.000457   0.000111   0.063841 (  0.719540)
curl on root with basic auth:  0.000389   0.000095   0.048539 (  0.807732)
curl on root with session auth:  0.000479   0.000116   0.186354 (  2.672784)
curl on multiple endpoints with session auth:  0.000859   0.000209   0.480265 ( 10.841015)
curl on multiple endpoints with basic auth:  0.000726   0.000177   0.367814 (  8.859796)
```

```
              user     system      total        real
curl on root without login:  0.000427   0.000096   0.057136 (  0.687104)
curl on root with basic auth:  0.000467   0.000104   0.048592 (  0.862687)
curl on root with session auth:  0.000421   0.000094   0.178445 (  2.568848)
curl on multiple endpoints with session auth:  0.001025   0.000229   0.576333 (  8.948681)
curl on multiple endpoints with basic auth:  0.000000   0.001359   0.450287 ( 10.347542)
```

```
              user     system      total        real
curl on root without login:  0.000448   0.000073   0.054231 (  0.678578)
curl on root with basic auth:  0.000406   0.000065   0.049842 (  0.867248)
curl on root with session auth:  0.000477   0.000078   0.178306 (  2.403250)
curl on multiple endpoints with session auth:  0.000971   0.000158   0.518806 (  9.558119)
curl on multiple endpoints with basic auth:  0.001447   0.000000   0.405758 ( 10.668350)
```

I ran it multiple times and results are similar to these ones : reaching 10 Redfish endpoints took between 8s and 12s using session or basic auth.

**Comment by neomilium** [comment]
> [...] One advantage of this approach is that it doesn't patch the RedfishClient class. That gem also looks unmaintained these days, so getting a proper patch in to support choosing behavior will be hard.

[It seems that a PR could be merged if suggested](https://github.com/xlab-steampunk/redfish-client-ruby/issues/7#issuecomment-2126294876)

File: modules/bmc/bmc_api.rb [hunk 1]
```diff
@@ -120,12 +117,11 @@ class Api < ::Sinatra::Base
       if params[:action].nil?
         return { :actions => ["status"] }.to_json
       end
-      bmc_setup
+      bmc_setup do |bmc|
       # determine which function should be executed
```
Comment (evgeni) [inline modules/bmc/bmc_api.rb:121]:
```suggestion
        # determine which function should be executed
```

**Comment by ekohl** [comment]
I had a hard time reading those results so here they are aligned:
```
                                               user       system     total       real
curl on root without login:                    0.000457   0.000111   0.063841 (  0.719540)
curl on root with basic auth:                  0.000389   0.000095   0.048539 (  0.807732)
curl on root with session auth:                0.000479   0.000116   0.186354 (  2.672784)
curl on multiple endpoints with session auth:  0.000859   0.000209   0.480265 ( 10.841015)
curl on multiple endpoints with basic auth:    0.000726   0.000177   0.367814 (  8.859796)

                                               user       system     total       real
curl on root without login:                    0.000427   0.000096   0.057136 (  0.687104)
curl on root with basic auth:                  0.000467   0.000104   0.048592 (  0.862687)
curl on root with session auth:                0.000421   0.000094   0.178445 (  2.568848)
curl on multiple endpoints with session auth:  0.001025   0.000229   0.576333 (  8.948681)
curl on multiple endpoints with basic auth:    0.000000   0.001359   0.450287 ( 10.347542)

                                               user       system      total      real
curl on root without login:                    0.000448   0.000073   0.054231 (  0.678578)
curl on root with basic auth:                  0.000406   0.000065   0.049842 (  0.867248)
curl on root with session auth:                0.000477   0.000078   0.178306 (  2.403250)
curl on multiple endpoints with session auth:  0.000971   0.000158   0.518806 (  9.558119)
curl on multiple endpoints with basic auth:    0.001447   0.000000   0.405758 ( 10.668350)
```

Looks like there's significant overhead to set up a session for a single request, but with 10 requests the result is inconclusive on what's better.

Given that information, I'm leaning to always using basic auth because it's simpler.

**Comment by ekohl** [comment]
I've opened https://github.com/xlab-steampunk/redfish-client-ruby/pull/8

**Comment by ekohl** [comment]
Since testing showed that there's no benefit to using sessions and 0.6.0 has a proper parameter to disable sessions I'm closing this.

**Pull request closed without merging**